File: ios/Classes/SmartLocalNotificationPlugin.swift
import Flutter
import UIKit
import UserNotifications
import AVFoundation

public class SmartLocalNotificationPlugin: NSObject, FlutterPlugin {
    private var audioManager: AudioManager?
    private var notificationManager: NotificationManager?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "smart_local_notification", binaryMessenger: registrar.messenger())
        let instance = SmartLocalNotificationPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    override init() {
        super.init()
        self.audioManager = AudioManager()
        self.notificationManager = NotificationManager()
        setupNotificationCategories()
    }

    private func setupNotificationCategories() {
        notificationManager?.initialize()
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            initialize(result: result)
        case "showNotification":
            showNotification(call: call, result: result)
        case "cancelNotification":
            cancelNotification(call: call, result: result)
        case "cancelAllNotifications":
            cancelAllNotifications(result: result)
        case "stopAudio":
            stopAudio(result: result)
        case "isAudioPlaying":
            isAudioPlaying(result: result)
        case "requestPermissions":
            requestPermissions(result: result)
        case "arePermissionsGranted":
            arePermissionsGranted(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func initialize(result: @escaping FlutterResult) {
        // Initialize audio session and notification center
        audioManager?.initialize()
        notificationManager?.initialize()
        result(true)
    }
    
    private func showNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        // Parse notification data and show notification with audio
        notificationManager?.showNotification(args: args) { [weak self] success in
            if success, let audioSettings = args["audioSettings"] as? [String: Any] {
                self?.audioManager?.playAudio(settings: audioSettings)
            }
            result(success)
        }
    }
    
    private func cancelNotification(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let id = args["id"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        notificationManager?.cancelNotification(id: id)
        result(true)
    }
    
    private func cancelAllNotifications(result: @escaping FlutterResult) {
        notificationManager?.cancelAllNotifications()
        result(true)
    }
    
    private func stopAudio(result: @escaping FlutterResult) {
        audioManager?.stopAudio()
        result(true)
    }
    
    private func isAudioPlaying(result: @escaping FlutterResult) {
        let isPlaying = audioManager?.isPlaying ?? false
        result(isPlaying)
    }
    
    private func requestPermissions(result: @escaping FlutterResult) {
        notificationManager?.requestPermissions { granted in
            result(granted)
        }
    }
    
    private func arePermissionsGranted(result: @escaping FlutterResult) {
        notificationManager?.arePermissionsGranted { granted in
            result(granted)
        }
    }
}

File: android/src/main/kotlin/com/example/smart_local_notification/SmartLocalNotificationPlugin.kt
package com.example.smart_local_notification

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** SmartLocalNotificationPlugin */
class SmartLocalNotificationPlugin: FlutterPlugin, MethodCallHandler, ActivityAware {
  private lateinit var channel: MethodChannel
  private lateinit var context: Context
  private var notificationManager: SmartNotificationManager? = null
  private var audioManager: SmartAudioManager? = null
  private var permissionManager: SmartPermissionManager? = null

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "smart_local_notification")
    channel.setMethodCallHandler(this)
    context = flutterPluginBinding.applicationContext
    
    // Initialize managers
    notificationManager = SmartNotificationManager(context)
    audioManager = SmartAudioManager(context)
    permissionManager = SmartPermissionManager(context)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    when (call.method) {
      "initialize" -> initialize(result)
      "showNotification" -> showNotification(call, result)
      "cancelNotification" -> cancelNotification(call, result)
      "cancelAllNotifications" -> cancelAllNotifications(result)
      "stopAudio" -> stopAudio(result)
      "isAudioPlaying" -> isAudioPlaying(result)
      "requestPermissions" -> requestPermissions(result)
      "arePermissionsGranted" -> arePermissionsGranted(result)
      else -> result.notImplemented()
    }
  }

  private fun initialize(result: Result) {
    try {
      notificationManager?.initialize()
      audioManager?.initialize()
      result.success(true)
    } catch (e: Exception) {
      result.success(false)
    }
  }

  private fun showNotification(call: MethodCall, result: Result) {
    try {
      val args = call.arguments as? Map<String, Any> ?: run {
        result.error("INVALID_ARGUMENTS", "Invalid arguments provided", null)
        return
      }

      val id = args["id"] as? Int ?: run {
        result.error("MISSING_ID", "Notification ID is required", null)
        return
      }

      val title = args["title"] as? String ?: ""
      val body = args["body"] as? String ?: ""
      val notificationSettings = args["notificationSettings"] as? Map<String, Any> ?: emptyMap()
      val audioSettings = args["audioSettings"] as? Map<String, Any>

      // Show notification
      val notificationShown = notificationManager?.showNotification(
        id, title, body, notificationSettings
      ) ?: false

      // Play audio if settings provided
      if (audioSettings != null && notificationShown) {
        audioManager?.playAudio(audioSettings)
      }

      result.success(notificationShown)
    } catch (e: Exception) {
      result.error("SHOW_NOTIFICATION_ERROR", e.message, null)
    }
  }

  private fun cancelNotification(call: MethodCall, result: Result) {
    try {
      val args = call.arguments as? Map<String, Any> ?: run {
        result.error("INVALID_ARGUMENTS", "Invalid arguments provided", null)
        return
      }

      val id = args["id"] as? Int ?: run {
        result.error("MISSING_ID", "Notification ID is required", null)
        return
      }

      notificationManager?.cancelNotification(id)
      result.success(true)
    } catch (e: Exception) {
      result.error("CANCEL_NOTIFICATION_ERROR", e.message, null)
    }
  }

  private fun cancelAllNotifications(result: Result) {
    try {
      notificationManager?.cancelAllNotifications()
      result.success(true)
    } catch (e: Exception) {
      result.error("CANCEL_ALL_NOTIFICATIONS_ERROR", e.message, null)
    }
  }

  private fun stopAudio(result: Result) {
    try {
      audioManager?.stopAudio()
      result.success(true)
    } catch (e: Exception) {
      result.error("STOP_AUDIO_ERROR", e.message, null)
    }
  }

  private fun isAudioPlaying(result: Result) {
    try {
      val isPlaying = audioManager?.isPlaying() ?: false
      result.success(isPlaying)
    } catch (e: Exception) {
      result.success(false)
    }
  }

  private fun requestPermissions(result: Result) {
    try {
      val granted = permissionManager?.requestPermissions() ?: false
      result.success(granted)
    } catch (e: Exception) {
      result.success(false)
    }
  }

  private fun arePermissionsGranted(result: Result) {
    try {
      val granted = permissionManager?.arePermissionsGranted() ?: false
      result.success(granted)
    } catch (e: Exception) {
      result.success(false)
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
    audioManager?.cleanup()
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    permissionManager?.setActivity(binding.activity)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    // No action needed
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    permissionManager?.setActivity(binding.activity)
  }

  override fun onDetachedFromActivity() {
    permissionManager?.setActivity(null)
  }
}
